File: Programmers/Level 2/조이스틱/Solution.java
// 푸는데 좀 걸림
// 가장 가까운 A가 아닌 문자에 먼저 접근하는 방식
// 왼쪽으로 갔을 때와 오른쪽으로 갔을 때의 거리 중 가까운 거리 사용
// 문자열 길이가 최대 20밖에 안되서 단순하게 모든 위치에서 시작한 후 도출된 최소 결과를 답으로 제출했다.

class Solution {
    static boolean[] visited;
    static int answer;
    static int answer_tmp;

    public static int solution(String name) {
        answer = Integer.MAX_VALUE;

        for (int i = 0; i < name.length(); i++) {
            answer_tmp = 0;
            visited = new boolean[name.length()];
            answer_tmp += Math.min(i, name.length() - (i + 1) + 1);

            int idx = moveNearest(name, i);

            while (idx != Integer.MAX_VALUE) {
                answer_tmp += Math.min(name.charAt(idx) - 'A', 'Z' - name.charAt(idx) + 1);

                idx = moveNearest(name, idx);
            }
            answer = Math.min(answer, answer_tmp);
        }
        return answer;
    }

    public static int moveNearest(String name, int n) {
        int idx = Integer.MAX_VALUE;
        int dist = Integer.MAX_VALUE;

        for (int i = 0; i < name.length(); i++) {
            if (!visited[i] && name.charAt(i) != 'A') {
                if (n < i) {
                    int tmp = Math.min(i - n, name.length() - i + n);
                    if (tmp < dist) {
                        dist = tmp;
                        idx = i;
                    }
                } else if (n > i) {
                    int tmp = Math.min(n - i, name.length() - n + i);
                    if (tmp < dist) {
                        dist = tmp;
                        idx = i;
                    }
                } else {
                    idx = i;
                    dist = 0;
                }
            }
        }

        if (idx != Integer.MAX_VALUE) {
            visited[idx] = true;
            answer_tmp += dist;
        }

        return idx;
    }

    public static void main(String[] args) {
        String name = "BAAAAABBB";
        System.out.println(solution(name));
    }
}

File: Programmers/Level 2/조이스틱/solution.kt
// 완전탐색 문제
// 예전에 풀었었는데 다시 푸니까 기억이 안남...
// 그냥, 새로운 방법을 찾았다.

// 상,하 움직임은 A부터 빠른지 Z부터 빠른지 간단하게 구하면된다.
// 따라서, 이 문제의 핵심은 좌,우 움직임의 최솟값을 어떻게 구하느냐이다.
// 가장 간단한 방법은 현재 위치까지 앞에서부터 오는게 빠른지, 뒤에서부터 시작하는게 빠른지를 비교하여 계속 최솟값을 갱신하는 방법이다.
// 이렇게 비교하는 이유는 'BBBBAAAAB'와 같은 경우 때문에 뒤부터 시작하는 것을 고려한다.
// 'A'가 연속될 경우가 존재하기 때문에 A 연속 이후의 index를 구한다.
// 앞에서부터 차례대로 i까지 오는 경우는 2 * i + (N - index)
// 뒤부터 시작하는 경우는 2 * (N - index) + i 이다.

var N = 0

fun solution(name: String): Int {
    var answer = 0
    var move = Int.MAX_VALUE
    N = name.length

    for (i in 0 until N) {
        answer += minOf(name[i] - 'A', 'Z' - name[i] + 1)

        var index = i + 1
        while (index < N && name[index] == 'A') index++

        move = minOf(move, 2 * i + (N - index), 2 * (N - index) + i)
    }

    return answer + move
}

fun main() {
    println(solution("BBBBAAAAB"))
}